### PR TITLE
Replicaset

### PR DESCRIPTION
**Pull request by rmohr**
Document the VirtualMachineReplicaSet

### CODE REVIEW

File: replicaset.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+### How to use a VirtualMachineReplicaSet
+
+The *VirtualMachineReplicaSet* allows to specify a *VirtualMachineTemplate* in
+`spec.template`. It consists of `ObjectMetadata` in `spec.template.metadata`, and
+a `VirtualMachineSpec` in `spec.template.spec`. The specification of the virtual
+machine is equal to the specification of the virtual machine in the
+`VirtualMachine` workload.
+
+`spec.replicas` can be used to specify how much replicas are wanted. If not
```
Comment (davidvossel) [inline replicaset.md]: s/much/many

File: replicaset.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+specified, the default value is 1. This value can be updated anytime. The
+controller will react to the changes.
+
+`spec.selector` is used by the controller to keep track of managed virtual
+machines. The selector specified there must be able to match the virtual
+machine labels as specified in `spec.template.metadata.labels`. If the selector
```
Comment (davidvossel) [inline replicaset.md]: This is a little strange from a user perspective. I would expect the VMRS to manage this label for us unless the user explicitly defines the label themselves.

File: replicaset.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+does not match these labels, or they are empty, the controller will simply do
+nothing except from logging an error. The user is responsible for not creating
+other virtual machines or *VirtualMachineReplicaSets* which conflict with the
+selector and the template labels.
+
+### When to use a VirtualMachineReplicaSet
+
+Whenever one wants many equal Virtual Machines and does not care about keeping
```
Comment (davidvossel) [inline replicaset.md]: s/equal/identical

File: replicaset.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+any kind of state, after they were terminated, the VirtualMachineReplicaSet is
```
Comment (davidvossel) [inline replicaset.md]: s/any kind of state, after they were terminated,/maintaining any disk state after they are terminated,/

File: replicaset.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+the right choice.
+
+#### Fast starting ephemeral Virtual Machines
+
+In this use-case one has small, fast booting Virtual Machines, and only minimal or
```
Comment (davidvossel) [inline replicaset.md]: I'd change this first sentence to something like the following. 

"This use-case involves small fast booting Virtual Machines with little provisioning performed during initialization."

File: replicaset.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+no provisioning is done during the boot. With other words, the timespan
+between the creation of a new Virtual Machine and entering the ready state of a
+VirtualMachine is short.
+
+Migrations are not important. Load balancing between nodes can be achieved by
```
Comment (davidvossel) [inline replicaset.md]: "Redistributing Virtual Machine workloads between"

File: replicaset.md [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+simply deleting managed VirtualMachines which are running on an overloaded
```
Comment (davidvossel) [inline replicaset.md]: simply by deleting

File: replicaset.md [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+Node. The `eviction` of such a VirtualMachine can happen by directly deleting
+the VirtualMachine instance (KubeVirt aware cluster load balancer) or by
+deleting the corresponding Pod where the Virtual Machine runs in (Only
+Kubernetes aware cluster load balancer).
```
Comment (davidvossel) [inline replicaset.md]: I know what you're talking about when you mention "cluster load balancer", but this term is going to be confusing for others. 

For example, if you google "kubernetes cluster load balancer"  the first results are going to talk about integrating k8s with external http load balancers and how the k8s service load balancer works. 

Workload distribution across the cluster is what we're referring to here. This segment probably needs to be restructured to better represent that.

File: replicaset.md [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+One still does not care about the state, but since re-provisioning
+VirtualMachines is expensive, Migrations are important. Load balancing between
+nodes can be achieved by migrating VirtualMachines to different nodes. A
+cluster load balancer needs to be aware of KubeVirt and create Migrations,
+instead of `evicting` VirtualMachines by deletion.
+
+> **Note:** The simplest form of having a migratable ephemeral VirtualMachine,
+> will be to use local storage based on `ContainerRegistryDisks` in combination
+> with a file based backing store. However, migratable backing store
```
Comment (davidvossel) [inline replicaset.md]: This is a good idea. I forgot it is technically possible to live migrate VMs using registry disks.

File: replicaset.md (hunk 1 above)
Comment (rmohr) [reply replicaset.md]: done

File: replicaset.md (hunk 2 above)
Comment (rmohr) [reply replicaset.md]: Yes it is not that nice. We can improve it a little by doing defaulting in both ways if possible. With the set based matchers it is not always possible to deduce labels for the VM.

File: replicaset.md (hunk 3 above)
Comment (rmohr) [reply replicaset.md]: done

File: replicaset.md (hunk 5 above)
Comment (rmohr) [reply replicaset.md]: done

File: replicaset.md (hunk 6 above)
Comment (rmohr) [reply replicaset.md]: done

File: replicaset.md (hunk 7 above)
Comment (rmohr) [reply replicaset.md]: done

File: replicaset.md (hunk 8 above)
Comment (rmohr) [reply replicaset.md]: done

File: replicaset.md [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+#### Slow starting ephemeral Virtual Machines
+
+In this use-case one has big, slow booting Virtual Machines, and complex or
+resource intensive provisioning is done during the boot. With other words, the
+timespan between the creation of a new Virtual Machine and entering the ready
+state of a VirtualMachine is long.
+
+One still does not care about the state, but since re-provisioning
+VirtualMachines is expensive, Migrations are important. Load balancing between
```
[the remark this answers is not in the document]
Comment (rmohr) [reply replicaset.md]: done

**Comment by davidvossel** [comment]
@rmohr I'd merge this, but I don't have write access to this repo. 

As soon as this is merged, I'll submit a PR to follow up on the ContainerRegistryDisk -> RegistryDisk change.

**Pull request merged**